File: src/main/scala/SlidingWindowMaximum.java
public class SlidingWindowMaximum {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if(nums == null || k == 0) return new int[0];
        int [] res = new int[nums.length - k + 1];
        java.util.PriorityQueue<Integer> queue = new java.util.PriorityQueue<>(new java.util.Comparator<Integer>(){
            public int compare(Integer o1, Integer o2) {
                return o2 - o1;
            }
        });
        for(int i = 0;i < k;i++){
            queue.add(nums[i]);
        }
        res[0] = queue.peek();
        for(int i = k;i< nums.length;i++){
            queue.add(nums[i]);
            queue.remove(nums[i - k]);
            res[i - k + 1] = queue.peek();
        }
        return res;
    }
}

File: src/main/scala/SlidingWindowMaximum.py
class Solution:
    def maxSlidingWindow(self, nums, k):
        n = len(nums)
        dq = []
        res = []
        for i in range(n):
            while dq and dq[0][0] <= i - k:
                dq.pop(0)
            while dq and dq[-1][1] <= nums[i]:
                dq.pop()
            dq.append([i, nums[i]])
            if i + 1 >= k:
                res.append(dq[0][1])
        return res


sol = Solution()
print(sol.maxSlidingWindow([1,3,-1,-3,5,3,6,7], 3))#[3,3,3,5,5,6,7]
print(sol.maxSlidingWindow([1,3,-1,-3,5,3,6,7], 2))#[3,3,-1,5,5,6,7]


